### ISSUE TITLE
Iterators of Proxy not very reliable

### ISSUE THREAD

**Issue by Kaiepi** [rakudo/rakudo#5035]
## The Problem

I'm chipping away at a more featureful implementation of [annotations](https://gist.github.com/Kaiepi/6e61d0402b6606f479e393a92888c9d7), but I find myself spending a lot of time fighting `Seq`-related operations WRT containers, which have also been a problem while working on `Data::Record` itself in other ways, because they tend to decont at inconvenient times. This is an (incomplete!) collection of several smaller issues:

```raku
class Scalarish is Proxy {
    has $!value;

    method new is raw {
        callwith :&FETCH, :&STORE
    }

    my method FETCH() {
        $!value
    }

    my method STORE(Mu $value is raw --> Nil) {
        $!value = $value
    }

    method multiply($times) {
        self.new xx $times
    }

    method sequence($a, $b) {
        (self.new = $a)...(self.new = $b)
    }

    method loop-once {
        do loop (my int $i; $i < 1; $i++) { self.new }
    }

    method reduce(&f, +scalarish) is raw {
        scalarish.reduce: &f
    }
}

proto itemized($ --> Bool:D) {*}
multi itemized(Mu $value is raw) {
    use nqp;
    ?nqp::istype_nd($value.VAR, Scalarish:D)
}
multi itemized(@values is raw) {
    cache @values.map: &itemized
}

put itemized Scalarish.new;
put itemized Scalarish.multiply(2);
put itemized Scalarish.sequence(1, 2);
put itemized Scalarish.loop-once;
put itemized Scalarish.reduce(-> \a, \b { b }, (Scalarish.new = 1), (Scalarish.new = 2));
put itemized (Scalarish.new = 1) Z= (Scalarish.new = 2);
```

## Expected Behavior
Should output a bunch of `True`.

## Actual Behavior
Outputs `True` once and then a bunch of `False`:
```
LAPTOP-7I9RCT0C:~/Documents/ra-annotations% raku golf.raku
True
False False
False False
False
False
False
```

## Steps to Reproduce

Save the given script as `golf.raku` and run it.

## Environment
* Operating system:
```
LAPTOP-7I9RCT0C:~/Documents/ra-annotations% uname -a
Linux LAPTOP-7I9RCT0C 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 GNU/Linux
```
* Compiler version (`perl6 -v` or `raku -v`):
```
Welcome to Rakudo™ v2022.07-13-g10feda116.
Implementing the Raku® Programming Language v6.d.
Built on MoarVM version 2022.07-2-g5e3ea9ce8.
```

**Comment by lizmat** [comment]
Are we sure that `Scalarish.new` is actually returning the `Proxy` ?

**Comment by Kaiepi** [comment]
It should be. `STORE`'s assignment can be wrapped with `once`, allowing this to output `True` and the rest of the output to remain as-is:
```raku
put 1 == ((Scalarish.new = 1) = 2);
```

**Comment by vrurg** [comment]
The problem is that not about iterators but about the operators and statements care about container's value, not the container itself. In particular, try replacing `multiply` body with something like `self.new, self.new` – you'll get double-`True`, as expected.

`...` is totally and undoubtedly has to have the values to operate, no surprise here.

With regard to the rest, I'm not ready to say if they have to pass over containers or values.

**Comment by Kaiepi** [comment]
> The problem is that not about iterators but about the operators and statements care about container's value, not the container itself. In particular, try replacing multiply body with something like self.new, self.new – you'll get double-True, as expected.

Yes, that should've been more clear. `map` and `for` preserve containers, so I've been riding off that so far.

**Comment by Kaiepi** [comment]
I know that the `reduce` case in particular is a bit complicated, and is what spurred this issue in particular, because it depends on `&take` internally. `&take-rw` is not something that can be swapped in without punishing users that don't depend on this functionality performance-wise. Most of the others relate to `Iterator`s from `Rakudo::Iterator` lacking `is raw` on their `pull-one` in some way; the C-loop case I was able to infer based off this.

**Comment by fingolfin0** [comment]
It looks like the problem I found is related to this one, so I thought I'd post an example here describing the malfunction of `for` with the rw-attribute:
```raku
class A does Iterable {
	has @!attr handles 'iterator' = 1..5;
}

class WithRoAttr {
	has A $.a .= new;
}
.say for WithRoAttr.new.a;     # Correct output: «1␤2␤3␤4␤5␤»
WithRoAttr.new.a.iterator.push-all(my @arr1);
.say for @arr1;                # Correct output: «1␤2␤3␤4␤5␤»

class WithRwAttr {
	has A $.a is rw .= new;
}
.say for WithRwAttr.new.a;       # Wrong output: «A.new»
WithRwAttr.new.a.iterator.push-all(my @arr2);
.say for @arr2;                # Correct output: «1␤2␤3␤4␤5␤»
```
